pkg/engine/lifecycletest/step_generator_test.go

```diff
@@ -366,3 +373,702 @@ func TestReadNilOutputs(t *testing.T) {
 				"BAIL: step executor errored: step application failed: resource 'my-resource-id' does not exist")
 		})
 }
+
+//nolint:paralleltest // This cannot be run in parallel because it depends on a global variable
+func TestConcurrentRegisterResource(t *testing.T) {
+	testCore := func() int64 {
+		load := atomic.Int64{}
+		maxLoad := atomic.Int64{}
+
+		loaders := []*deploytest.ProviderLoader{
+			deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+				return &deploytest.Provider{
+					CheckF: func(urn resource.URN, olds, news resource.PropertyMap, randomSeed []byte) (
+						resource.PropertyMap, []plugin.CheckFailure, error,
+					) {
+						currLoad := load.Add(1)
+						defer load.Add(-1)
+
+						for currMax := maxLoad.Load(); currLoad > currMax; currMax = maxLoad.Load() {
+							maxLoad.CompareAndSwap(currMax, currLoad)
+						}
+
+						// Sleeping in tests is a recipe for flakiness, but without a delay, there is no
+						// opportunity for concurrency
+						time.Sleep(time.Millisecond * 5)
+						runtime.Gosched()
+
+						return news, nil, nil
+					},
+				}, nil
+			}),
+		}
+
+		const resourceCount = 10
+
+		programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			// create resources concurrently
+			wg := sync.WaitGroup{}
+			for i := 0; i != resourceCount; i++ {
+				index := i
+
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+
+					name := fmt.Sprintf("test-%c", index+'a')
+					_, err := monitor.RegisterResource("pkgA:m:typA", name, true)
+					assert.NoError(t, err)
+				}()
+			}
+
+			wg.Wait()
+
+			return nil
+		})
+		hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+		p := &TestPlan{
+			Options: TestUpdateOptions{
+				HostF:                      hostF,
+				OnlyVerifyCompleteSnapshot: true,
+			},
+		}
+
+		project := p.GetProject()
+		snap, err := TestOp(Update).Run(project, p.GetTarget(t, nil), p.Options, false, p.BackendClient, nil)
+		assert.NoError(t, err)
+
+		assert.NotNil(t, snap)
+		if snap != nil {
+			actualResUrns := make([]resource.URN, resourceCount)
+			expectedResUrns := make([]resource.URN, resourceCount)
+			for i, res := range snap.Resources[1 : 1+resourceCount] {
+				actualResUrns[i] = res.URN
+				expectedResUrns[i] = resource.URN(fmt.Sprintf("urn:pulumi:test::test::pkgA:m:typA::test-%c", 'a'+i))
+			}
+
+			assert.ElementsMatch(t, expectedResUrns, actualResUrns)
+		}
+
+		return maxLoad.Load()
+	}
+
+	//nolint:paralleltest // This cannot be run in parallel because it depends on a global variable
+	t.Run("Serialized Step Generator", func(t *testing.T) {
+		deploy.UseParallelStepGen = false
```

Comment:
We should probably also reset this variable after the test, otherwise `--parallel-step-gen` is a bit unpredictable.

Reply:
good call, i'll reset this in all tests where I set it